### DISCUSSION TITLE
Multiselection of blocks (how useful/necessary is it?)

### DISCUSSION THREAD

**Discussion by BeksOmega** [RaspberryPiFoundation/blockly-keyboard-experimentation#9] [General]
Is it important for folks to be able to interact with multiple blocks at once? If it is important, do we need support for non-contiguous blocks, or only contiguous blocks?

**Comment by kmcnaught** [comment]
I think this is definitely important for the equivalent of dragging a whole stack of blocks, e.g. from one event handler to another. I don't think it needs to include non-contiguous blocks.

**Reply by cpcallen** [reply]
It's worth remembering that when using the mouse to drag a block all child blocks, including any blocks attached to the next connection, are moved too.  I don't know the full history but I suspect that this was orignally simply expedience due to the way the underlying block tree and associated SVG (and XML!) is structured, but it sufficiently obviated the need for multiselect that we've not done that yet, and it and probably persists in large part _because of_ the lack of multiselect.

I think multiselect (contiguous and noncontiguous—e.g., multiple stacks) is a useful feature in its own right, but implementing it will give us an opportunity to revisit the dragging behaviour.  That's tricky, though, because moving and rearranging stacks by grabbing their top block is a pretty core Blockly mechanic, and changing this would break a lot of instructional material even if it provided a better overall experience (though I'm not sure it would: moving a very tall stack of blocks (not enclosed in an event block) would be tricky even with multiselect).

We don't necessarily need to make keyboard-based block movement work the same way as mouse dragging but it's not clear that there's a compelling reason to make it work differently (and, until we do have multiselect, there's a compelling reason not to).

**Reply by cpcallen** [reply]
It appears to me that if this feature is "definitely important" then you are expecting that keyboard-based block movement work differently than mouse dragging currently works (i.e., it would not move the remainder of the stack below the block being moved)—and if so I would like to understand the reasons why you think it desirable to make keyboard- and mouse-driven block movement behaviour different.

**Reply by kmcnaught** [reply]
I don't think it's necessarily essential to make keyboard- and mouse-driven block movement exactly equivalent, since there are a number of ways users might want to move blocks, some of which may be better suited to one input method or the other. But for each type of movement UX, if they feel semantically equivalent then I agree we should try to keep the behaviour as similar as possible between mouse and keyboard. 

### Dragging

We think that a "dragging" style operation will be possible with the proposed 2d cartesian movement. You'd detach a block or blocks, move around the canvas, and snap it to the new location. Whilst it's important to have this UX available for positioning top-level blocks, I don't think this is likely to be the most efficient/easy way to move blocks when using alternate input controls. 

It makes sense to aim for mouse and keyboard dragging to be equivalent, so I think I'd be happy for this operation to act on a stack in the first instance for consistency. We'll see if that has any unexpected UX downsides. I guess this means that "detach" always means "detach this and children". I hadn't necessarily pictured this, but I'm also not immediately against it. 

I don't currently have strong opinions on this user flow, other than _"I want it to be possible but avoidable as much as possible"_!. 

### Cut/Copy/Paste

If we have mouse and keyboard focus integrated correctly, then we would hope that a mouse user could 
- Click on the source block, press Ctrl+C. 
- Click on a block next to the destination, 
- press arrow up/down to find connection, 
- press Ctrl+V to paste. 

A keyboard user could do the same with keyboard nav to get to the source + destination. 

As I detailed in this discussion about [different behaviour across different editors](https://github.com/google/blockly-keyboard-experimentation/discussions/10#discussioncomment-10866339), Cut/Copy/Paste in MakeCode already acts only on a single block, whereas in Scratch they _mainly_ operate on stacks. So there isn't a single pre-existing behaviour to copy here. I'd prefer single block (or explicit multi-selection) to be the default, and let platforms override the behaviour if they want something different. 

### Cut/Copy/Paste of multiple blocks

For a keyboard user, you would:
- navigate to the source location
- hold Shift and press arrow(s) to multi select
- Cut or Copy
- move to destination
- Paste

This would also be useful for a mouse user, giving them more control than the current mouse drag (which applies to the whole stack) or the current copy behaviour (which applies either to 1 block or a whole stack, depending on editor). They would:

- Click on first source block. 
- Hold shift and click on the last source block (the resulted multiselection would be constrained to the current scope)
- Copy
- Either:
  - Click near the destination block, press arrow up/down to reach connection
  - Paste
- Or: (this just came to me as a consideration)
  - Click on the canvas away from blocks
  - Paste to have it appear detached on the canvas
  - Drag it to the destination
  
### Moving blocks inline

We've also discussed moving block(s) up/down a line, without detaching them. I think this would be useful for mouse and keyboard users. For example:

Keyboard user:
- Move to source block
- Optionally use Shift + arrows to multiselect
- Hold a modifier (Alt?) or press a key to toggle a mode
- Use arrow keys to move selection up/down within current stack

Mouse user:
- Click to select source block
- Optionally use Shift + click another block to multiselect
- Hold a modifier (Alt?) or press a key to toggle a mode
- Use arrow keys to move selection up/down within current stack

We haven't discussed the details of this UX flow for a while, but we are imagining something like this would exist - there may be some details to iron out. It may be that this is only appropriate for blocks with a prev/next connection, and not blocks with an in/out connection (since the latter would have more nuanced constraints on suitable placement which might be hard to reason about / communicate)

### AOB

Am I missing other methods of moving blocks? Where does your perspective / vision differ from what I've detailed above? 

In terms of moving a massive stack of blocks, are we still envisaging that eventually there will be shortcuts for "go to end of scope" ? If so, in an IDE you'd expect the following to be possible:
- navigate to start of source stack
- hold shift
- press the shortcut to go to end of scope
- Now you've multi-line selected the entire stack

Whether or not this would be possible in Blockly would depend how we implemented the shortcuts, since it requires the 2 shortcuts are compatible, and it might be veering too far into "making an adult IDE". But just wanted to raise it as a possible mitigation.
